### PR TITLE
Quick pass on the writing exporters docs

### PR DESCRIPTION
**Pull request by jamtur01**
Emphasis on clarity, consistency, and tone.

### CODE REVIEW

**Comment by jamtur01** [comment]
Also, where do folks stand with regard to the Oxford comma?

File: content/docs/instrumenting/writing_exporters.md [hunk 1]
```diff
@@ -5,67 +5,46 @@ sort_rank: 5
...
 If the system in question has only a handful of metrics that rarely change,
-then getting everything perfect is an easy choice (e.g. the [haproxy
-exporter](https://github.com/prometheus/haproxy_exporter)).
+then getting everything perfect is an easy choice, a good example of this is the [HAProxy exporter](https://github.com/prometheus/haproxy_exporter).
 
-If on the other hand the system has hundreds of metrics that change
-continuously with new versions, if you try to get things perfect then you’ve
-signed yourself up for a lot of ongoing work. The [mysql
-exporter](https://github.com/prometheus/mysqld_exporter) is on this end of the
+If, on the other hand, the system has hundreds of metrics that change
```
Comment (matthiasr) [inline content/docs/instrumenting/writing_exporters.md]: I'd rephrase this a little more to get rid of the double if, like:

> On the other hand, if you try to get things perfect when the system has hundreds of metrics that change frequently with new versions, then you’ve signed yourself up for a lot of ongoing work.

File: content/docs/instrumenting/writing_exporters.md [hunk 2]
```diff
@@ -5,67 +5,46 @@ sort_rank: 5
...
 
 ## Configuration
 
 When working with applications, you should aim for an exporter that requires no
 custom configuration by the user beyond telling it where the application is.
 You may also need to offer the ability to filter out certain metrics if they
-may be too granular and expensive on large setups (e.g. the haproxy exporter
-allows filtering of per-server stats). Similarly there may be expensive metrics
-that are disabled by default.
+may be too granular and expensive on large setups, for example the [HAProxy exporter](https://github.com/prometheus/haproxy_exporter) allows filtering of per-server stats. Similarly there may be expensive metrics that are disabled by default.
```
Comment (matthiasr) [inline content/docs/instrumenting/writing_exporters.md]: The line breaks are very inconsistent now.

File: content/docs/instrumenting/writing_exporters.md [hunk 3]
```diff
@@ -316,83 +244,64 @@ Rather create new metrics each time. In Go this is done with
...
 
 Note that it is only currently possible to write this type of exporter with the
 Go, Python and Java client libraries.
 
-The other is where you’re pulling some stats out of a random instance of a
-system and don’t care which one you’re talking to. Consider a set of MySQL
-slaves you wanted to run some business queries against the data to then export.
-Having an exporter that uses your usual load balancing approach to talk to one
-slave is the sanest approach.
+The second exception is where you’re pulling some stats out of a random instance of a system and don’t care which one you’re talking to. Consider a set of MySQL slaves you wanted to run some business queries against the data to then export. Having an exporter that uses your usual load balancing approach to talk to one slave is the sanest approach.
```
Comment (matthiasr) [inline content/docs/instrumenting/writing_exporters.md]: While we're at it, WDYT about fixing slavery metaphors here? I think "replica" is a perfectly self-explanatory alternative.

File: content/docs/instrumenting/writing_exporters.md [hunk 4]
```diff
@@ -401,42 +310,40 @@ Metrics should only be pulled from the application when Prometheus scrapes
...
+Some applications and monitoring systems only push metrics, for example StatsD,
+Graphite and collectd.
 
-There’s two considerations here.
+There are two considerations here.
 
 Firstly, when do you expire metrics? Collected and things talking to Graphite
 both export regularly, and when they stop we want to stop exposing the metrics.
-Collected includes an expiry time so we use that, Graphite doesn’t so it’s a
+Collected includes an expiry time so we use that, Graphite doesn’t so it is a
```
Comment (matthiasr) [inline content/docs/instrumenting/writing_exporters.md]: I think this should be `Collectd`?

File: content/docs/instrumenting/writing_exporters.md [hunk 5]
```diff
@@ -90,17 +67,16 @@ Metric names for applications should generally be prefixed by the exporter
 name, e.g. `haproxy_up`.
 
 Metrics must use base units (e.g. seconds, bytes) and leave converting them to
-something more readable to the graphing software. No matter what units you end
+something more readable to upstream tools. No matter what units you end
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: I meant graphing software, this is a frontend concern.

File: content/docs/instrumenting/writing_exporters.md [hunk 6]
```diff
@@ -90,17 +67,16 @@ Metric names for applications should generally be prefixed by the exporter
...
 up using, the units in the metric name must match the units in use. Similarly
-expose ratios, not percentages (though a counter for each of the two components
-of the ratio is better).
+expose ratios, not percentages. Even better, specify a counter for each of the two components of the ratio.
 
-Metric names should not include the labels that they’re exported with (e.g.
-`by_type`) as that won’t make sense if the label is aggregated away.
+Metric names should not include the labels that they’re exported with, e.g.
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: I don't think this is as clear

File: content/docs/instrumenting/writing_exporters.md [hunk 7]
```diff
@@ -109,7 +85,7 @@ Prometheus metrics and label names are written in `snake_case`. Converting
 doesn’t always produce nice results for things like `myTCPExample` or `isNaN`
 so sometimes it’s best to leave them as-is.
 
-Exposed metrics should not contain colons, these are for users to use when
+Exposed metrics should not contain colons, these are reserved for use when
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: It needs to be clear that these are for end-user use. "aggregating" has a few different meanings depending on who you're talking to.

File: content/docs/instrumenting/writing_exporters.md [hunk 8]
```diff
@@ -123,40 +99,23 @@ suffixes.
...
-the pid, both Go and Python offer collectors that’ll handle this for you (see
-the [haproxy exporter](https://github.com/prometheus/haproxy_exporter) for an
-example).
+prefix on to these if they follow the [matching semantics](https://docs.google.com/document/d/1Q0MXWdwp1mdXCzNRak6bW5LLVylVRXhdi7_21Sg15xQ/edit). For example, Prometheus has `scrape_duration_seconds` for how long a scrape took, it’s good practice to also have an exporter-centric metric, e.g. `jmx_scrape_duration_seconds`, saying how long the specific exporter took to do its thing. For process stats where you have access to the PID, both Go and Python offer collectors that’ll handle this for you. A good example of this is the [HAProxy exporter](https://github.com/prometheus/haproxy_exporter).
 
 When you have a successful request count and a failed request count, the best
 way to expose this is as one metric for total requests and another metric for
 failed requests. This makes it easy to calculate the failure ratio. Do not use
-one metric with a failed/success label. Similarly with hit/miss for caches,
+one metric with a failed or success label. Similarly with hit/miss for caches,
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: Be consistent between 'or' and '/'

File: content/docs/instrumenting/writing_exporters.md (hunk 1 above)
Comment (jamtur01) [reply content/docs/instrumenting/writing_exporters.md]: Yes - thanks much nicer.

File: content/docs/instrumenting/writing_exporters.md (hunk 2 above)
Comment (jamtur01) [reply content/docs/instrumenting/writing_exporters.md]: I'm going to fix the whole file.

File: content/docs/instrumenting/writing_exporters.md (hunk 3 above)
Comment (jamtur01) [reply content/docs/instrumenting/writing_exporters.md]: Awesome - wasn't sure - some projects are receptive to this - others it's a 500 comment storm. Will fix.

File: content/docs/instrumenting/writing_exporters.md (hunk 4 above)
Comment (jamtur01) [reply content/docs/instrumenting/writing_exporters.md]: Yes. Fixed.

File: content/docs/instrumenting/writing_exporters.md [hunk 9]
```diff
@@ -229,256 +255,277 @@ metrics, stop.
...
+types. This usually means counters and gauges. The `_count` and `_sum`
+of summaries are also relatively common, and on occasion you’ll see
+quantiles. Histograms are rare, if you come across one remember that the
+exposition format exposes cumulative values.
 
-Often it won’t be obvious what the type of a metric is (especially if you’re
-automatically processing a set of metrics), use `UNTYPED` in that case. In
-general `UNTYPED` is a safe default.
+Often it won’t be obvious what the type of metric is, especially if
+you’re automatically processing a set of metrics, use `UNTYPED` in that
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: I think end the sentence at "metrics" might be a little clearer.

File: content/docs/instrumenting/writing_exporters.md (hunk 3 above)
Comment (brian-brazil) [reply content/docs/instrumenting/writing_exporters.md]: In this case replica is actually more accurate.

File: content/docs/instrumenting/writing_exporters.md (hunk 9 above)
Comment (jamtur01) [reply content/docs/instrumenting/writing_exporters.md]: Yeah. It's a bit repetitive.

File: content/docs/instrumenting/writing_exporters.md [hunk 10]
```diff
@@ -5,179 +5,202 @@ sort_rank: 5
...
+In other cases, metrics from the system are completely non-standard,
+depending on the usage of the system and the underlying application.  In
+that case the user has to tell us how to transform the metrics. The [JMX
+exporter](https://github.com/prometheus/jmx_exporter) is the worst
+offender here, with the
+[Graphite](https://github.com/prometheus/graphite_exporter) and
+[StatsD](https://github.com/prometheus/statsd_exporter) exporters also
+requiring configuration to extract labels.
+
+Providing a configuration that produces some output automatically and a
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: If configuration is provided then it's not automatic

File: content/docs/instrumenting/writing_exporters.md [hunk 11]
```diff
@@ -5,179 +5,202 @@ sort_rank: 5
...
+selection of example configurations for transformation, if required, is
+advised. Here are some guidelines for writing or adapting custom
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: Last sentence here doesn't fit.

File: content/docs/instrumenting/writing_exporters.md [hunk 12]
```diff
@@ -5,179 +5,202 @@ sort_rank: 5
...
-The rule of thumb is that one metric should make sense when summed or averaged.
-There is one other case that comes up with exporters, and that’s where the data
-is fundamentally tabular and doing otherwise would require users to do regexes
-on metric names to be usable. Consider the voltage sensors on your
-motherboard, while doing math across them is meaningless, it makes sense to
-have them in one metric rather than having one metric per sensor. All values
-within a metrics should (almost) always have the same unit (consider if fan
-speeds were mixed in with the voltages, and you had no way to automatically
-separate them).
+Read or write and send or receive are best as separate metrics, rather than as
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: Read and write is a set here, so 'or' doesn't make sense.

File: content/docs/instrumenting/writing_exporters.md [hunk 13]
```diff
@@ -195,32 +218,35 @@ my_metric{label=b} 6
...
-full query pattern and is sufficient for uniqueness. However it is of little
-use without the human readable `digest_text` label, which for long queries will
-contain only the start of the query pattern and is thus not unique. Thus we end
-up with both the `digest_text` for humans and the `digest` label for
-uniqueness.
+The former breaks for people who do a `sum()` over your metric, and the
+latter breaks sum and is quite difficult to work with. Some client
+libraries, for example Go, will actively try to stop you doing the
+latter in a custom collector, and all client libraries should stop you
+from doing the former with direct instrumentation. Never do either of
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: s/former/latter/

File: content/docs/instrumenting/writing_exporters.md [hunk 14]
```diff
@@ -221,8 +222,8 @@ my_metric{label=b} 6
 The former breaks for people who do a `sum()` over your metric, and the
 latter breaks sum and is quite difficult to work with. Some client
 libraries, for example Go, will actively try to stop you doing the
-latter in a custom collector, and all client libraries should stop you
-from doing the former with direct instrumentation. Never do either of
+former in a custom collector, and all client libraries should stop you
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: They both should be "latter". There's no way we can really stop the former.

File: content/docs/instrumenting/writing_exporters.md [hunk 15]
```diff
@@ -74,14 +74,15 @@ offender here, with the
...
-advised. Here are some guidelines for writing or adapting custom
-exporters.
+Ensuring the exporter works out of the box without configuration, and
+providing a selection of example configurations for transformation if
+required, is advised.
 
 YAML is the standard Prometheus configuration format, all configuration
 should use YAML by default.
 
+Here are some guidelines for writing or adapting custom exporters.
```
Comment (brian-brazil) [inline content/docs/instrumenting/writing_exporters.md]: This sentence should be removed, this is not an overview section.

**Comment by brian-brazil** [comment]
Thanks!
